src/wovnio/html/HtmlConverter.php

```diff
@@ -276,10 +276,31 @@ private function insertHreflangTags($html)
         return $this->insertAfterTag($parent_tags, $html, implode('', $hreflangTags));
     }
 
+    private function translateCanonicalTag($html)
+    {
+        if ($this->isNoindexLang($this->headers->requestLang())) {
+            return $html;
+        }
+
+        $canonical_tag_regex = "/(<link[^>]*rel=\"canonical\"*[^>]*href=\")([^\"]*)(\"\>)/";
+        preg_match($canonical_tag_regex, $html, $matches);
+        if (count($matches) < 4) {
+            return $html;
+        }
+        $original_canonical_url = $matches[2];
+        $translated_canonical_url = $this->convertUrlToLanguage($original_canonical_url, $this->headers->requestLang());
+        $replacement = '\1' . $translated_canonical_url . '\3';
```

Comment:
basically the idea is to use 3 capture groups with $2 being the canonical URL we want to replace. `\1` and `\3` is just PHP's way of doing back-reference.